horovod/common/operations.cc

```diff
@@ -1001,51 +1001,58 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
                               DDL_OP_SUM))
 #else
       if (horovod_global.hierarchical_allreduce) {
-        NCCL_CHECK(entries, "ncclReduce",
-                   ncclReduce(fused_input_data, buffer_data,
-                              (size_t)num_elements,
-                              GetNCCLDataType(first_entry.tensor), ncclSum, 0,
+        int64_t num_elements_per_gpu = num_elements%horovod_global.local_size == 0 ?
```

Comment:
What if we do float32, `num_elements=8`, `buffer_len=32` and we have `local_size=6`?
* `num_elements_per_gpu` will be `2`.
* `buffer_len_per_proc` will be `32/8*2=8`, which will run beyond the buffer in `ncclAllGather()`.

We need to make sure buffer is appropriately padded.

Reply:
I have padded `buffer_data` and `fused_input_data` before the hierarchical allreduce so that the buffer sizes are a multiple of `local_size`.